includes/subscriptions/class-wc-payments-subscriptions-event-handler.php

```diff
@@ -175,12 +200,25 @@ public function handle_invoice_payment_failed( array $body ) {
 		$event_data            = $this->get_event_property( $body, 'data' );
 		$event_object          = $this->get_event_property( $event_data, 'object' );
 		$wcpay_subscription_id = $this->get_event_property( $event_object, 'subscription' );
-		$wcpay_invoice_id      = $this->get_event_property( $event_object, 'id' );
-		$attempts              = (int) $this->get_event_property( $event_object, 'attempt_count' );
-		$subscription          = WC_Payments_Subscription_Service::get_subscription_from_wcpay_subscription_id( $wcpay_subscription_id );
+
+		/**
+		 * When a store is in staging mode, we don't want any webhook handling response to be sent to the server.
+		 *
+		 * Sending requests from staging sites can have unintended consequences for the live store. For example,
+		 * Subscriptions which renew on the staging site will lead to paused subscriptions at Stripe and result in
+		 * missed renewal payments.
+		 */
+		if ( WCS_Staging::is_duplicate_site() ) {
+			$this->log_skipped_webhook_due_to_staging( 'invoice.payment_failed', $wcpay_subscription_id );
+			return;
+		}
+
+		$wcpay_invoice_id = $this->get_event_property( $event_object, 'id' );
+		$attempts         = (int) $this->get_event_property( $event_object, 'attempt_count' );
+		$subscription     = WC_Payments_Subscription_Service::get_subscription_from_wcpay_subscription_id( $wcpay_subscription_id );
 
 		if ( ! $subscription ) {
-			throw new Invalid_Webhook_Data_Exception( __( 'Cannot find subscription for the incoming "invoice.upcoming" event.', 'woocommerce-payments' ) );
```

Comment:
This was a copy paste fail in the existing code. This function handles `invoice.payment_failed` events, not `invoice.upcoming`.